cli/snapshot.go

```diff
@@ -0,0 +1,169 @@
+package cli
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// ioctl magics from <linux/fs.h>
+	IOCTL_FIFREEZE_MAGIC = 0xC0045877 // _IOWR('X', 119, int)
+	IOCTL_FITHAW_MAGIC   = 0xC0045878 // _IOWR('X', 120, int)
+)
+
+type SnapshotOptions struct {
+	ArtifactName      string
+	ArtifactGroup     string
+	CompatibleDevices []string
+	Depends           []string
+	Provides          []string
+}
+
+type mountInfo struct {
+	mountID    int
+	mountPoint string
+	mountSrc   string
+	fsType     string
+}
+
+func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
+	rootDev, err := getFsDevFile("/")
+	if err != nil {
+		return err
+	}
+	dataDev, err := getFsDevFile(runOpts.dataStore)
+	if err != nil {
+		return err
+	}
+	if rootDev == dataDev {
+		return errors.Errorf(
+			"State data store (%s) is located on rootfs partition",
+			runOpts.dataStore)
+	}
+
+	errChan := make(chan error)
+	go func() {
```

Comment:
Why the outer `go func()`? It seems to just return an error at the end, so it's pretty synchronous.